### PR TITLE
bumping recycler view version to 23.2.1 for Android, definition of It…

### PR DESCRIPTION
**Pull request by naqvitalha**
Most of the existing android apps use recycler view versions **23.2.1 and above**. ReactAndroid still depends on **23.0.1**. In the new versions of recycler view the definition of **RecyclerView.ItemAnimator** abstract class has changed and since gradle would choose the latest build of recycler view the apps would start to get following unhandled exception on runtime: 

`java.lang.AbstractMethodError: abstract method "boolean android.support.v7.widget.RecyclerView$ItemAnimator.animatePersistence(android.support.v7.widget.RecyclerView$ViewHolder, android.support.v7.widget.RecyclerView$ItemAnimator$ItemHolderInfo, android.support.v7.widget.RecyclerView$ItemAnimator$ItemHolderInfo)"
                                                                          at android.support.v7.widget.RecyclerView$4.processPersistent(RecyclerView.java:464)
                                                                          at android.support.v7.widget.ViewInfoStore.process(ViewInfoStore.java:243)
                                                                          at android.support.v7.widget.RecyclerView.dispatchLayoutStep3(RecyclerView.java:3330)
                                                                          at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3080)
                                                                          at android.support.v7.widget.RecyclerView.consumePendingUpdateOperations(RecyclerView.java:1505)
                                                                          at android.support.v7.widget.RecyclerView.access$400(RecyclerView.java:151)
                                                                          at android.support.v7.widget.RecyclerView$1.run(RecyclerView.java:305)
                                                                          at android.view.Choreographer$CallbackRecord.run(Choreographer.java:858)
                                                                          at android.view.Choreographer.doCallbacks(Choreographer.java:670)
                                                                          at android.view.Choreographer.doFrame(Choreographer.java:603)
                                                                          at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:844)
                                                                          at android.os.Handler.handleCallback(Handler.java:739)
                                                                          at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                          at android.os.Looper.loop(Looper.java:148)
                                                                          at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                          at java.lang.reflect.Method.invoke(Native Method)
                                                                          at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                          at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
`

This is exactly what happened while I tried to do the same on one of my existing apps.

To workaround this and make integration with existing apps seamless I've upgraded recycler view version and written new implementation for **NotAnimatedItemAnimator** class.

I've tested with this change and new versions (23.2.1+) of the library work just fine now. The new implementation is almost similar to the old one. Only method names have changed in new interface.

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @cjhopman and @mkonicek to be potential reviewers.

**Comment by facebook-github-bot-5** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-9** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

File: ReactAndroid/src/main/java/com/facebook/react/views/recyclerview/NotAnimatedItemAnimator.java [hunk 1]
```diff
@@ -10,49 +12,38 @@
 /*package*/ class NotAnimatedItemAnimator extends RecyclerView.ItemAnimator {
 
   @Override
-  public void runPendingAnimations() {
-    // nothing
+  public boolean animateDisappearance(@NonNull RecyclerView.ViewHolder viewHolder, @NonNull ItemHolderInfo preLayoutInfo, @Nullable ItemHolderInfo postLayoutInfo) {
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/views/recyclerview/NotAnimatedItemAnimator.java]: Please follow the style conventions, see the original file

**Comment by facebook-github-bot-5** [comment]
@naqvitalha updated the pull request.

File: ReactAndroid/src/main/java/com/facebook/react/views/recyclerview/NotAnimatedItemAnimator.java (hunk 1 above)
Comment (naqvitalha) [inline ReactAndroid/src/main/java/com/facebook/react/views/recyclerview/NotAnimatedItemAnimator.java]: Done

**Comment by ide** [comment]
CircleCI tests are failing and need to be fixed: https://circleci.com/gh/facebook/react-native/10096 http://pastebin.com/MkGjhh2Z

**Comment by naqvitalha** [comment]
@ide I see a BUCK file inside react-native/ReactAndroid/src/main/third-party/android/support/v7/recyclerview, It was pointing to the wrong remote file

`android_prebuilt_aar(
    name = 'recyclerview',
    aar = ':recyclerview-binary-aar',
    visibility = [
        '//ReactAndroid/...',
    ],
)

remote_file(
    name = 'recyclerview-binary-aar',
    url = 'mvn:com.android.support:recyclerview-v7:aar:23.0.1’,  //this
    sha1 = '94d5f16be156521e135295e2ffa70d79ef8ad9d5',
)`

Will changing this to the right version work? Don't have much idea about BUCK

**Comment by satya164** [comment]
cc @bestander @astreet

**Comment by bestander** [comment]
Bumping to Android SDK is a hassle.

The problem is that at FB we have a monorepo for all mobile apps and they share all the dependencies. This protects us from a great pain of dealing with multiple versions of the same library between apps and frameworks.
But it comes at a price of bigger effort to update third party libraries, especially the core ones like Android SDK because they are used everywhere.

This is true for React Native. The positive part is that we can maintain high quality of master branch of React Native because we test it across many apps without a delay.

As for the recyclerview, we can update to 23.1.0 because actually we are already using it internally.
As for 23.2.1, I would rather wait because it is an effort to update everything at once, is there any strong reason to do it now?

If anyone wants to introduce and abstraction that will allow any version of recyclerview for React Native, please let me know, we could coordinate efforts.
I understand it can be frustrating for people who integrate RN into existing apps.

**Comment by bestander** [comment]
@naqvitalha, yes, you need to change the remote file in
```
remote_file(
name = 'recyclerview-binary-aar',
url = 'mvn:com.android.support:recyclerview-v7:aar:23.0.1’, //this
sha1 = '94d5f16be156521e135295e2ffa70d79ef8ad9d5',
)
```

**Comment by facebook-github-bot-5** [comment]
@naqvitalha updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@naqvitalha updated the pull request.

**Comment by naqvitalha** [comment]
@bestander 23.1.0 also contains the new interface definition so that'll work too. I've switched to 23.1.0, let's see if it passes CircleCI this time

**Comment by bestander** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-4** [comment]
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3709326) to review internal test results.

**Comment by bestander** [comment]
Sorry, it gets a bit delayed.
There is some complication in the build scripts that make it harder and is not as straight forward as I thought.
If you think this is really really needed for your app, please comment here and I'll have a look sooner.
Otherwise I need to jump into some high priority issues now.

**Comment by duranmla** [comment]
I got the same issue, perhaps, I was about to check if I can make a PR to be able to bumping the recyclerView version to the most recent one (which I think is `24.1.1` according to: https://developer.android.com/topic/libraries/support-library/features.html) this need happens to me cause I wanted to integrate the last version of Intercom within my application but the recyclerView version from react native and intercom collide and throw errors.

**Comment by naqvitalha** [comment]
This is very important in my opinion since this makes react native incompatible with almost every existing app out there using a recycler view.  Not using recyclerviewbackedscrollview is not an option due to all the weird animation glitches like list scrolling up on its own.  I've already got it working on my end,  I raised the PR to save other people some time

**Comment by ide** [comment]
> the weird animation glitches like list scrolling up on its own

fyi this has been fixed

**Comment by naqvitalha** [comment]
In which version?  Still happens in 0.30

**Comment by ide** [comment]
0.32 I think?

**Comment by duranmla** [comment]
So @ide what will be the approach in order to make integrations with sdk like intercom's that use recyclerView 24 once the 0.32 is released?

**Comment by ide** [comment]
I'm not sure really. If v24's public API is similar enough to v23.1.0's, after @bestander figures how to merge this PR, you probably could set up Gradle (somehow...) to use v24 and RN would compile fine. Alternatively you could fork RN and make it use v24.

**Comment by bestander** [comment]
The reason why the merge fails is because many FB apps that use React Native still use 23.0.1.
To merge the PR I would need to go through the apps and try to bump them all to 23.1.0.

Technically appcompat 23.1.0 is already available there but some apps are still behind.
Eventually we will update them and it does not seem like a large job but I don't have time to do that in the next couple of weeks, unfortunately.

There are only 4 Java files that depend on appcompat.v7 libraries
https://github.com/facebook/react-native/search?l=java&q=%22android.support.v7%22&type=Code&utf8=%E2%9C%93.

Can anyone motivated just abstract that into an interface so that people could provide their own implementation of the version of the appcompat if needed?
I'll help out with Buck but can't lead this now.
This could be a good start of decoupling the version of Android SDK from RN.

**Comment by lacker** [comment]
FWIW it looks like people concluded that shimming out the RecyclerView would be too hard in https://github.com/facebook/react-native/pull/9583 . So it seems like the only way forward is to figure out how to bump the version number in RN itself & just get our internal tests passing. I am going to import to try running internal tests on this again.

**Comment by lacker** [comment]
Looks like this is just a massively breaking change if you have apps that are part-React-Native-hybrids that depend on the old version of RecyclerView, which is what basically all the FB apps are. I'm not really sure how to get around that. It does seem bothersome to be stuck on this old version forever, but if the pain of merely upgrading the FB apps past this breaking change is a proxy for how pained the community would be in net, I am hesitant. So for now I am sort of stalled in despair. Maybe we should just wait until someone invents a way to have this change be less breaky.

**Comment by lacker** [comment]
It seems like this is too hard to fix for now so I am going to close this pull request. Sorry. If anyone has a way to make progress here then I encourage opening a new one or objecting.

**Comment by jpshelley** [comment]
I don't think the best solution is to close the Issue? How long is Facebook planning on staying on `23.0.1`? You'll want to update eventually, correct? This is somewhat frustrating for those of us that are currently on `25.0.1`.

Check out the Android documentation
> Provided an easy transition plan for this backward-incompatible API change:
If you’ve previously extended the ItemAnimator class, you can change your base class to SimpleItemAnimator and your code should work as before. The SimpleItemAnimator class provides the old API by wrapping the new API.

https://developer.android.com/topic/libraries/support-library/revisions.html

This should be easy enough of a change to get RN corrected yes?

**Comment by lacker** [comment]
I just don't want to imply that this pull request might get merged in by leaving it open. For now we are not accepting updates to the version here because it causes a lot of breaking with the internal apps we use to test things. I think we are likely to revisit in the future but not via this pull request. If there is some non-breaking way to migrate this, we could do that, but it doesn't seem like there is one.

**Comment by jpshelley** [comment]
@lacker Thats fine as long as the conversation stays open and transparent. I don't think there is a non-breaking way to migrate either, just need to rip the band-aid. Doesn't it make more sense for React Native to update and put the pressure back on to Facebook to update their own applications? React Native is an OSS and should strive to support the community especially if others are willing to contribute to it and submit pull requests.

Instead of going back and forth though about what is best for Facebook+Community, would it be feasible for Facebook to create an `ANDROID-UPDATE-25.0.1.md` page in the RN repo to let the OSS community know the plans and dates for expected updates as well as ways where we can help?

**Comment by satya164** [comment]
Facebook is always in master. It's not possible to update it without breaking Facebook apps. So they have to be updated first.

**Comment by jpshelley** [comment]
For those interested I have a fork with the updates for `25.0.1` here https://github.com/jpshelley/react-native/tree/0.39.0-AndroidSupportLibrary25.0.1

I'm trying to get jitpack to work with it so it can be easily added to a project. I won't have the time to keep it up to date constantly, but will try until React Native updates on their Repo. This should allow others to use React Native `0.39.0` with the most up to date support libraries.

(You can also host on your own locally if I don't end up getting jitpack to work, we use Artifactory internally and its working great)

**Comment by bestander** [comment]
We are in the process of updating Android Support library to 23.4.0.
Can't promise on the date but stay tuned.

**Pull request closed without merging**